Add hash and int128 utils for Lazy Tensor Core

wconstab opened the pull request:
These utils are prerequisites for Lazy Node base class.

- set up new torch/csrc/lazy, test/cpp/lazy dirs
- add source files to build_variables.bzl in new lazy_core_sources var
- create new test_lazy binary

Fixes #65636

JackCaoG commented on torch/csrc/lazy/core/hash.cpp:
```diff
@@ -0,0 +1,81 @@
+    hash_t h;
+    std::array<uint8_t, sizeof(hash_t)> b;
+  } uval = {.h = 0};
+  std::memcpy(uval.b.data(), *data, size);
```
I think this part is from pt/xla, but I am curious if you know why do we need to memcpy to `uval.b.data()` here?

wconstab replied:
I don't see the necessity of doing it this way.  Clearly the aim is to copy a 128-bit chunk from an arbitrary point in a uint8_t buffer, and then return it as a hash_t.

But why bother with a union and an explicit memcpy?  Couldn't we cast *data to a hash_t* and use that to initialize a hash_t for return?  (And not need the union type at all?)

It's possible someone did it this way to work around some kind of linter requirement.

davidel replied:
The `memcpy()` is required to deal with architectures which do not allow unaligned memory access (or they pay big penalty).
Note that these days a `memcpy()` maps to a compiler builtin, and the compiler uses the most appropriate code depending on the architecture.
Example, on x86_64 this is just a move:

https://godbolt.org/z/4escG8WWE

The union is needed to read the last partial bytes, padding the missing ones with zeros.

JackCaoG commented on torch/csrc/lazy/core/int128.h:
```diff
@@ -0,0 +1,138 @@
+namespace torch {
+namespace lazy {
+
+class alignas(unsigned __int128) uint128 {
```
Out of curiosity why not use `absl::` version of int128 utils and type?

ezyang replied:
Yeah, definitely something to consider. The int128 here was taken straight from XLA, right? It doesn't seem like a very good implementation; probably should consider cribbing a more fully featured one from somewhere.

wconstab replied:
I think this version was made (simplified) by @asuhan during the process of making lazy_tensor_core separate from XLA (and absl) dependencies.  What would it entail for PyTorch to use the absl 128 type? Is it a light enough dep that we would just add entire absl as a thirdparty dep to PyTorch?  Or are you proposing to copy just int128 from absl, and thereby get a more full featured one.

ezyang replied:
copy it in. Also not sure if absl is the best lib to get 128-bit from; it might be a pass if it's very intricately tied to absl

wconstab replied:
Apart from it being a 1K LOC file and needing relatively sparse changes to remove ABSL stuff, it has other `.inc` files I'd have to sort out which define int128 intrinsics, which I'm not sure we'd even use?
https://github.com/abseil/abseil-cpp/blob/master/absl/numeric/int128.h

```
// TODO(absl-team): This module is inconsistent as many inline `uint128` methods
// are defined in this file, while many inline `int128` methods are defined in
// the `int128_*_intrinsic.inc` files.
```

I could work on copying the absl 128 in if that's the preferred route.  I saw the chromium project also using absl's 128 and didn't find a lot of alternatives,

This could work too, but it's not a ton smaller than absl and I'd trust absl over it.
https://github.com/calccrypto/uint128_t/blob/master/uint128_t.include

wconstab replied:
ok, so leave the basic int128 in lazy/core, or still move to c10/util? @ezyang

ezyang replied:
up to you. It's still pretty reasonable to put it in c10/util on principle, but I guess if we think this implementation is crappy keeping it outside can prevent other people from starting to use it. The real problem is if someone else adds an int128 implementation without telling you

JackCaoG commented on test/cpp/lazy/test_misc.cpp:
```diff
@@ -0,0 +1,25 @@
+
+TEST(HashTest, Sanity) {
+    // repeatable
+    auto str_a = std::string("Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut at suscipit purus.");
```
Ideally we also want to test types like `c10::ScalarType` where we explicitly support, but I realized pt/xla never had those.... so not a hard requirement 😄

wconstab replied:
Yea, good point- i'll add coverage for a lot of types.

JackCaoG reviewed (commented):
Thanks! some minor questions.

ezyang commented on torch/csrc/lazy/core/int128.h:
```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <cstdint>
+#include <limits>
```
BTW, this file I'd be pretty comfortable putting in c10/utils, it's pretty uncontroversial (unlike the hash function)

ezyang reviewed (approved):
I'm not going to block landing this, esp if @JackCaoG gives the all clear. However, as you've mentioned, I think most of us don't really understand the current hash algorithm. I think it would be pretty helpful to have a more detailed discussion of what the important properties of the hash for lazy tensor should be, so that it is clearer what we can replace it with. In particular, is it important that this hash is collision free in practice? It feels like this is what is being attempted here (esp because the hash is extended to 128-bit) but it would be very helpful for it to be explicitly spelled out here.

wconstab commented:
> I think most of us don't really understand the current hash algorithm.

Yes, this is a risk and concern. I believe the best person to comment on the history of this (in particular, where this algo came from, and why it was important to use 128 v 64) is @dlibenzi, who isn't actively working on the project anymore.  

If we can get more context it would be great to document it, and in the future I think we could also do more focused experiments on whether we can replace this hash with a standard library.  But I don't want to do that experimentation until we have a stable system with good test coverage.

alanwaketan commented on c10/util/int128.h:
```diff
@@ -0,0 +1,143 @@
+/*
+ * A very basic uint128 class, based on absl version, used as the basis for
+ *128-bit hashing for the lazy tensor project. If using this for other purposes,
+ *it might make sense to replace this with a more complete implementation.
```
Nit: A space is needed after *.

wconstab replied:
lol.  really?  I had to delete those spaces to pass the linter (!)

alanwaketan replied:
Then, never mind.

alanwaketan reviewed (approved):
LGTM.

alanwaketan commented:
@wconstab Notice this is targeting master. Wondering if you will need to make another change in the lazy_tensor_staging branch to rebase/adopt this change.

wconstab commented:
> @wconstab Notice this is targeting master. Wondering if you will need to make another change in the lazy_tensor_staging branch to rebase/adopt this change.

Yea, after merging to master, I'd then merge master into ltc_staging_branch and in the merge commit I'd also delete the original hash in ltc_staging and update any includepaths.  Does this make sense (as a model workflow for our future PRs)?

alanwaketan commented:
> > @wconstab Notice this is targeting master. Wondering if you will need to make another change in the lazy_tensor_staging branch to rebase/adopt this change.
> 
> Yea, after merging to master, I'd then merge master into ltc_staging_branch and in the merge commit I'd also delete the original hash in ltc_staging and update any includepaths. Does this make sense (as a model workflow for our future PRs)?

I think it makes a lot sense. I can share some of my merge master tips with you. One comment about, maybe you can do i) a PR about merging master to the staging branch, and ii) another PR to adopt the changes.

JackCaoG reviewed (approved):
Thanks @wconstab. Do you want me to start the migration for pt/xla or you will do it?

wconstab commented:
> Thanks @wconstab. Do you want me to start the migration for pt/xla or you will do it?

I'll do this one if you'll do the next one ;) 

I don't have torch/xla set up to build locally, though, so I might bug you about that tmrw.  I started making the expected changes.

nkreeger commented on c10/util/int128.h:
```diff
@@ -0,0 +1,388 @@
+  TORCH_API friend std::ostream& operator<<(std::ostream& o, const uint128& b);
+
+ private:
+  static void DivModImpl(
```
Might be the cause of your msvc compiler error here - with this function being static I wonder if the `<<=` op returning a reference wants a static version of that overloaded op.

wconstab commented on c10/util/int128.cpp:
```diff
@@ -0,0 +1,184 @@
+    int difference = dividend_bit_length - divisor_bit_length;
+    uint128 quotient = 0;
+    while (difference >= 0) {
+      quotient <<= 1;
```
@nkreeger I don't follow your point.  I see that DivModImpl is a static function.  `quotient <<= 1` should be invoking a non-static member of the quotient object though, right?  It wouldn't make logical sense to implement an operator<<= method as static, would it?  (I didn't try though..)

Also, what do you mean about an overload?

wconstab replied:
i went ahead and changed operator<<= from a method to a static method, which I still don't quite understand the reasoning behind, but we'll see if CI is happy.

wconstab committed:
Add hash and int128 utils for Lazy Tensor Core (#65635)

Summary:
These utils are prerequisites for Lazy Node base class.

- set up new torch/csrc/lazy, test/cpp/lazy dirs
- add source files to build_variables.bzl in new lazy_core_sources var
- create new test_lazy binary

Fixes https://github.com/pytorch/pytorch/issues/65636

Pull Request resolved: https://github.com/pytorch/pytorch/pull/65635

Reviewed By: alanwaketan

Differential Revision: D31260343

Pulled By: wconstab

fbshipit-source-id: 5d14ed321ad29e2920fd79b51989924f3c0add38

suo commented:
This PR broke master: https://hud.pytorch.org/commit/e94fea08d0590e131f9762b000216b44f8c892da. Sadly it broke a binary build that does not run on PRs. @zhouzhuojie it looks like ciflow does not have the option to run this build?
